tests/test_mup.py

```diff
@@ -0,0 +1,305 @@
+# Copyright Contributors to the Cellarium project.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import math
+from collections.abc import Callable
+from pathlib import Path
+from typing import Literal
+
+import pandas as pd
+import pytest
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torchvision import datasets, transforms
+
+from cellarium.ml.models import MuLinear
+from cellarium.ml.utilities.testing import assert_slope_equals, get_coord_data
+
+optim_dict = {"sgd": torch.optim.SGD, "adam": torch.optim.Adam, "adamw": torch.optim.AdamW}
+
+
+# adapted from https://github.com/microsoft/mup/blob/main/examples/MLP/main.py
+def coord_check_MLP(
+    mup: bool,
+    bias: bool,
+    nonlin: Callable[[torch.Tensor], torch.Tensor],
+    lr: float,
+    input_mult: float,
+    output_mult: float,
+    optim_name: Literal["sgd", "adam", "adamw"],
+    train_loader: torch.utils.data.DataLoader,
+    nsteps: int,
+    nseeds: int,
+    widths: list[int],
+) -> pd.DataFrame:
+    def gen(w: int) -> Callable[[], nn.Module]:
+        def f() -> nn.Module:
+            model: nn.Module
+            if mup:
+                model = MuMLP(
+                    width=w,
+                    bias=bias,
+                    nonlin=nonlin,
+                    optimizer=optim_name,
+                    input_mult=input_mult,
+                    output_mult=output_mult,
+                )
+            else:
+                model = MLP(width=w, bias=bias, nonlin=nonlin, input_mult=input_mult, output_mult=output_mult)
+            return model
+
+        return f
+
+    models = {w: gen(w) for w in widths}
+    optim_fn = optim_dict[optim_name]
+
+    return get_coord_data(
+        models,
+        train_loader,
+        loss_fn=F.cross_entropy,
+        lr=lr,
+        optim_fn=optim_fn,
+        nseeds=nseeds,
+        nsteps=nsteps,
+    )
+
+
+@pytest.fixture
+def train_loader(tmp_path: Path) -> torch.utils.data.DataLoader:
+    batch_size = 64
+    transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+    trainset = datasets.CIFAR10(root=tmp_path, train=True, download=True, transform=transform)
+    return torch.utils.data.DataLoader(trainset, batch_size=batch_size, shuffle=True)
+
+
+# SP MLP
+class MLP(nn.Module):
```

Comment:
Add docstring (e.g. A Multi-Layer Perceptron (MLP) in Standard Parametrization (SP).)